### PR TITLE
Use SpiDriver to setup ETH

### PR DESCRIPTION
**Pull request by reinismu**
This should allow to share ETH with other devices on the same SPI bus

### CODE REVIEW

File: src/eth.rs [hunk 1]
```diff
@@ -295,13 +295,9 @@ impl<'d> EthDriver<'d> {
     esp_idf_eth_spi_ethernet_ksz8851snl
```
Comment (ivmarkov) [inline src/eth.rs:295]: In general: does not build for ESP IDF 5. Please check

File: src/eth.rs [hunk 2]
```diff
@@ -357,7 +346,7 @@ impl<'d> EthDriver<'d> {
         ),
         EspError,
     > {
-        Self::init_spi_bus::<P>(sclk, sdo, sdi, dma)?;
+        unsafe { gpio_install_isr_service(0) };
```
Comment (ivmarkov) [inline src/eth.rs:349]: Why is this necessary?

File: src/eth.rs [hunk 3]
```diff
@@ -295,13 +295,9 @@ impl<'d> EthDriver<'d> {
     esp_idf_eth_spi_ethernet_ksz8851snl
 ))]
 impl<'d> EthDriver<'d> {
-    pub fn new_spi<P: spi::Spi + 'd>(
-        spi: impl Peripheral<P = P> + 'd,
+    pub fn new_spi(
+        spi_driver: &spi::SpiDriver<'d>,
```
Comment (ivmarkov) [inline src/eth.rs:299]: The SPI driver should be configured with DMA mode enabled and with transfer size > 1500. Shall we check that?

Comment (ivmarkov) [inline src/eth.rs:299]: You are re-using the `SpiDriver` thing, but you are not re-using `SpiDeviceDriver`. The latter would allow you to get rid of `init_spi_device` and `get_spi_conf`. Since this is an obtrusive change, why not bite the bullet and pass-in an `SpiDeviceDriver` instance instead?

**Review by ivmarkov — CHANGES_REQUESTED** [review]

File: src/eth.rs (hunk 2 above)
Comment (reinismu) [reply src/eth.rs:349]: It was there before. Not sure why, but I guess it suppresses this error `E (473) gpio: gpio_isr_handler_add(465): GPIO isr service is not installed, call gpio_install_isr_service() first`

File: src/eth.rs (hunk 3 above)
Comment (reinismu) [reply src/eth.rs:299]: Is it the case for all chips or just W5500?

File: src/eth.rs (hunk 2 above)
Comment (ivmarkov) [reply src/eth.rs:349]: Pretty sure all chips.

File: src/eth.rs (hunk 3 above)
Comment (ivmarkov) [reply src/eth.rs:299]: Pretty sure all chips.

Comment (reinismu) [reply src/eth.rs:299]: Plan was to just change eth.rs API. Passing in `SpiDeviceDriver` instead of `SpiDriver` would mean that caller need to do all the configuration on his own. Wouldn't want to expose that complexity.

I do see how we could create `SpiDeviceDriver` internally to get rid of `init_spi_device` and `get_spi_conf`, but that can be done later as well.

Comment (ivmarkov) [reply src/eth.rs:299]: What additional configuration do you mean specifically? Please point me to it.

As for this being done later - I'm not feeling comfortable breaking the API twice just because. So we take either the current  route (`SpiDriver`) or what I suggested (`SpiDeviceDriver`). With good arguments for one or the other.

Comment (reinismu) [reply src/eth.rs:299]: `max_transfer_size` is not exposed from `SpiDriver`. Any suggestions how I could assert it?

Comment (ivmarkov) [reply src/eth.rs:295]: Oh well. Forget it. Users will get a cryptic error message, they need to decypher. :) Exposing bits and pieces from the configuration of the `SpiDriver` just to assert it here does not feel OK.

File: src/eth.rs [hunk 4]
```diff
@@ -295,13 +291,9 @@ impl<'d> EthDriver<'d> {
...
```
Comment (ivmarkov) [inline src/eth.rs:295]: I think we are in a pretty good shape now. The only thing remaining I think which is a bit annoying is that we do not have the flexibility to pass the SPI driver either by reference, or by value (The latter is more convenient when you don't plan to share the SPI bus with multiple devices - see for reference `SpiDeviceDriver`; but of course means you pay the extra verbosity of having the `EthDriver` having a non-lifetime generic now.). Also, folks using `EthDriver` with the RMII and the `openeth` impls will also pay the price (I guess, for them the generic would be resolved as `()`).

Any thoughts?

**Comment by teamplayer3** [comment]
closes #176

File: src/eth.rs (hunk 4 above)
Comment (reinismu) [reply src/eth.rs:295]: Hmm ye not sure what would be best course of action here. What do you think?

Comment (reinismu) [reply src/eth.rs:295]: I looked into this again and even together with GPT-4 I couldn't find a way :)
Most likely will need to redo whole ETH from ground up to make it play nice

Comment (ivmarkov) [reply src/eth.rs:295]: I'll also try to look over the weekend at it. Ideally we should merge this before the upcoming release of the esp-idf-* crates.

Comment (ivmarkov) [reply src/eth.rs:295]: Weekend keeps postponing, but also we have not released either. Sorry for the delay - perhaps *this* weekend. I basically want to see if the `BorrowMut` pattern we use elsewhere can be ergonomic enough here as well. I think it can if we define
```rust
pub struct RMII {}
pub struct OpenEth {}
pub struct SPI<T> where T: BorrowMut<SpiDriver>`
```

and then we parameterize `EthDriver` by `EthDriver<'d, Q>` where Q: is one of `RMII`, `OpenEth` or `SPI<T>`. This might result in a small win (don't remember Rust's method overloading rules exactly) where we can rename back `new_spi`, `new_rmii`, `new_openeth` all to just `new` except that they would be implemented on different parameterizations of `Q`. Might or might not work, have to check.

Comment (reinismu) [reply src/eth.rs:295]: Maybe make common trait `EthDriver` for all of them? Then pass it in `EspEth`. Then have `OpenEthDriver`, `W5500Driver` and `RMIIDriver` that will implement it.

Comment (ivmarkov) [reply src/eth.rs:295]: This way we might have to duplicate a lot of code across the the three implementations, and I'm not sure how that's fixing the main issue: which is - when using SPI - provide the user the capability to either pass the SPI driver by reference, or move it into the driver, where the latter would be like 99% of the cases (where they don't need to share the SPI bus with something else).

Anyway - my approach actually works. Ahem, type-checks.
I'll be merging your patch shortly, and then committing my approach on top of it.
